File: arrays/01_flood_fill/test_flood_fill.cpp
#include "flood_fill.cpp"
#include <gtest/gtest.h>
#include <vector>
#include <chrono>
#include <iostream>
 
using namespace std;


// A simple struct to hold test results
struct TestResult {
    string test_name;
    chrono::duration<double> elapsed_time;
};

// Custom listener
class CustomTestListener : public ::testing::EmptyTestEventListener {
private:
    vector<TestResult> results;
    chrono::time_point<chrono::high_resolution_clock> start_time;

public:
    // Called before a test starts
    void OnTestStart(const ::testing::TestInfo& test_info) override {
        start_time = chrono::high_resolution_clock::now();
    }

    // Called after a test ends
    void OnTestEnd(const ::testing::TestInfo& test_info) override {
        auto end_time = chrono::high_resolution_clock::now();
        chrono::duration<double> elapsed_time = end_time - start_time;
        results.push_back({test_info.name(), elapsed_time});
    }

    // Called after all tests have run
    void OnTestProgramEnd(const ::testing::UnitTest& unit_test) override {
        // Now print all the results
        cout << "\nTest Time Results:\n\n";
        int nameWidth = 60;
        for (const auto& result : results) {
            cout << left << setw(nameWidth) << result.test_name 
                 << "time: " << scientific << setprecision(2) 
                 << result.elapsed_time.count() << " s\n";
        }
        cout << endl;
    }
};


class TestFloodFill : public ::testing::Test {
protected:
    Solution* solution; // Each test will have its own pointer to Solution instance

    void SetUp() override {
        // Code here will be called immediately after the constructor 
        // (right before each test).
        solution = new Solution();
    }

    void TearDown() override {
        // Code here will be called immediately after each test 
        // (right before the destructor).
        delete solution;
    }
};


TEST_F(TestFloodFill, test1_SimpleCase) { 
    vector<vector<int>> image = {{1, 1, 1}, {1, 1, 0}, {1, 0, 1}};
    int sr = 1, sc = 1, newColor = 2;
    vector<vector<int>> expected = {{2, 2, 2}, {2, 2, 0}, {2, 0, 1}};
    vector<vector<int>> result = solution->floodFill(image, sr, sc, newColor);

    ASSERT_EQ(expected, result);
}

TEST_F(TestFloodFill, test2_SameStartColor) { 
    vector<vector<int>> image = {{0, 1, 2}, {3, 4, 5}};
    int sr = 0, sc = 0, newColor = 0;
    vector<vector<int>> expected = {{0, 1, 2}, {3, 4, 5}};
    vector<vector<int>> result = solution->floodFill(image, sr, sc, newColor);

    ASSERT_EQ(expected, result);
}

TEST_F(TestFloodFill, test3_ComplexCase) { 
    vector<vector<int>> image = {{6, 5, 1}, {2, 1, 1}, {1, 1, 3}, {1, 0, 1}};
    int sr = 2, sc = 1, newColor = 7;
    vector<vector<int>> expected = {{6, 5, 7}, {2, 7, 7}, {7, 7, 3}, {7, 0, 1}};
    vector<vector<int>> result = solution->floodFill(image, sr, sc, newColor);

    ASSERT_EQ(expected, result);
}


int main(int argc, char **argv) {
    ::testing::InitGoogleTest(&argc, argv);

    // Adds the custom listener to the end. Google Test takes the ownership.
    ::testing::TestEventListeners& listeners = ::testing::UnitTest::GetInstance()->listeners();
    listeners.Append(new CustomTestListener);

    return RUN_ALL_TESTS();
}

File: arrays/01_flood_fill/test_flood_fill.py
from unittest import TestCase, TestResult, TestLoader
from flood_fill import Solution
from time import time
import numpy as np

from functools import wraps


class CustomTestResult(TestResult):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.test_times = []

    def startTest(self, test):
        self._start_time = time()
        super().startTest(test)

    def stopTest(self, test):
        elapsed_time = time() - self._start_time
        self.test_times.append((test.id().split('.')[-1], elapsed_time))
        super().stopTest(test)

    def printTestTimes(self):
        name_width = 60
        print("\nTest Time Results:\n")
        for name, elapsed_time in self.test_times:
            print(f"{name:<{name_width}}time: {elapsed_time:.2e} s")
        print()


def time_tests(cls):
    """Class decorator for adding time measurements to all test methods of a TestCase."""
    for attr in dir(cls):
        if attr.startswith('test'):
            old_method = getattr(cls, attr)

            @wraps(old_method)
            def new_method(self, *args, method=old_method, **kwargs):
                method(self, *args, **kwargs)
            setattr(cls, attr, new_method)
    return cls


class TestFloodFill(TestCase):
    def setUp(self):
        # This method is called before each test
        self.solution = Solution()

    def test1_SimpleCase(self):
        image = np.array([[1, 1, 1], [1, 1, 0], [1, 0, 1]])
        sr, sc, new_color = 1, 1, 2
        expected = np.array([[2, 2, 2], [2, 2, 0], [2, 0, 1]])
        result = self.solution.flood_fill(image, sr, sc, new_color)

        np.testing.assert_array_equal(result, expected)

    def test2_SameStartColor(self):
        image = np.array([[0, 1, 2], [3, 4, 5]])
        sr, sc, new_color = 0, 0, 0
        expected = np.array([[0, 1, 2], [3, 4, 5]])
        result = self.solution.flood_fill(image, sr, sc, new_color)

        np.testing.assert_array_equal(result, expected)

    def test3_ComplexCase(self):
        image = np.array([[6, 5, 1], [2, 1, 1], [1, 1, 3], [1, 0, 1]])
        sr, sc, new_color = 2, 1, 7
        expected = np.array([[6, 5, 7], [2, 7, 7], [7, 7, 3], [7, 0, 1]])
        result = self.solution.flood_fill(image, sr, sc, new_color)

        np.testing.assert_array_equal(result, expected)

    def tearDown(self):
        del self.solution  # Ensure the reference is deleted.


if __name__ == '__main__':
    suite = TestLoader().loadTestsFromTestCase(TestFloodFill)
    result = CustomTestResult()
    suite.run(result)
    result.printTestTimes()
